### ISSUE TITLE
shaky tests on travis

### ISSUE THREAD

**Issue by ralphtheninja** [airtap/airtap#152]
- [ ] Re-run tests until they fail
- [ ] Copy log into this issue

**Comment by vweevers** [comment]
Latest build failed on node 8: [log.txt](https://github.com/airtap/airtap/files/1997604/log.txt). Relevant bit:

```
events.js:183
      throw er; // Unhandled 'error' event
      ^
Error: watch /home/travis/build/airtap/airtap/package.json ENOSPC
    at _errnoException (util.js:1022:11)
    at FSWatcher.start (fs.js:1382:19)
    at Object.fs.watch (fs.js:1408:11)
    at createFsWatchInstance (/home/travis/build/airtap/airtap/node_modules/chokidar/lib/nodefs-handler.js:37:15)
    at setFsWatchListener (/home/travis/build/airtap/airtap/node_modules/chokidar/lib/nodefs-handler.js:80:15)
    at FSWatcher.NodeFsHandler._watchWithNodeFs (/home/travis/build/airtap/airtap/node_modules/chokidar/lib/nodefs-handler.js:228:14)
    at FSWatcher.NodeFsHandler._handleFile (/home/travis/build/airtap/airtap/node_modules/chokidar/lib/nodefs-handler.js:255:21)
    at FSWatcher.<anonymous> (/home/travis/build/airtap/airtap/node_modules/chokidar/lib/nodefs-handler.js:473:21)
    at FSReqWrap.oncomplete (fs.js:153:5)
npm ERR! Test failed.  See above for more details.
```

It's watching too many files. We should disable `watchify` on CI builds.

**Comment by ralphtheninja** [comment]
> It's watching too many files.

Why is it watching files at all and why are they too many? Maybe watching files should be explicit by `airtap --watch`? (I'm really curious why they are too many)

**Comment by vweevers** [comment]
Maybe it has something to do with Docker?

+1 to opt-in `--watch`.

**Comment by ralphtheninja** [comment]
Would be nice to debug this just to get some information on which files are being watched. Might be a bug as well.

**Comment by vweevers** [comment]
There are two instances of watchify. One is for user files:

https://github.com/airtap/airtap/blob/b25021d85b43c02c35b17d5c331cb7e71f596f6b/lib/builder-browserify.js#L118

Second is for our instrumentation:

https://github.com/airtap/airtap/blob/b25021d85b43c02c35b17d5c331cb7e71f596f6b/lib/control-app.js#L90

Neither is configured to ignore `node_modules`. For that we need to set [the `ignoreWatch` option to true](https://github.com/browserify/watchify#options).

**Comment by vweevers** [comment]
Got a different kind of failure on memdown: [memdown airtap log.txt](https://github.com/airtap/airtap/files/2036039/memdown.airtap.log.txt)

```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

**Comment by vweevers** [comment]
Ah, the Sauce Labs dashboard shows an error:

```
An unknown server-side error occurred while processing the command.

Original error: Unable to start WebDriverAgent session because of 
xcodebuild failure: "Could not proxy command to remote server.

Original error: Error: connect ECONNREFUSED 127.0.0.1:8100".
```

And on retry:

```
An unknown server-side error occurred while processing the command.

Original error: Unable to launch WebDriverAgent because of 
xcodebuild failure: "xcodebuild failed with code 65".
```

**Comment by ralphtheninja** [comment]
https://github.com/airtap/browsers/issues/3#issuecomment-368733561

**Comment by ralphtheninja** [comment]
> Yeah, it's fixed: admc/wd#522. We gotta activate greenkeeper on airtap ;)

Done.

**Comment by feross** [comment]
I've seen the "No output has been received in the last 10m0s, this potentially indicates a stalled build" error before but have no idea what causes it.

In any case, I just merged https://github.com/airtap/airtap/pull/165 which upgrades `wd` so at least the `jsonData.value` bug is fixed.

**Comment by ralphtheninja** [comment]
I'm wondering if this watching thing is related to the other problems. Assuming it's related to file descriptors it isn't _that_ far fetched.

**Comment by vweevers** [comment]
I wonder if we can somehow disable services on the Sauce Labs side that we don't need. Like xcodebuild, UiAUtomator (I don't know what either of these do).

**Comment by ralphtheninja** [comment]
> I wonder if we can somehow disable services on the Sauce Labs side that we don't need. Like xcodebuild, UiAUtomator (I don't know what either of these do).

What I'd really like to do is to meet with the Sauce Labs engineers (once the level case study is done).

**Comment by ralphtheninja** [comment]
Another problem that I haven't seen before

[log1.txt](https://github.com/airtap/airtap/files/2079911/log1.txt)

**Comment by ralphtheninja** [comment]
@vweevers If you take a look at https://github.com/airtap/airtap/issues/152#issuecomment-392011017 where `t.plan()` fails I don't really think it's an error in the tests. If you check the log, there are six browsers being tested (6 * 8 + 2 = 50 tape tests in total) and one of the browsers fail (android:6.0) but it is actually _restarted_ and the tests are ok after the restart:

```
2018-05-25T10:04:33.251Z airtap:sauce:android:6.0 open http://airtap.local:40573/airtap
2018-05-25T10:04:34.861Z airtap:control-app zuul client took 0.531 seconds to bundle
2018-05-25T10:04:35.753Z airtap:browserify test files took 0.115 seconds to bundle
2018-05-25T10:04:35.978Z airtap:sauce:android:6.0 browser opened url
2018-05-25T10:04:35.978Z airtap:sauce:android:6.0 quitting browser
2018-05-25T10:04:38.390Z airtap:sauce:android:6.0 shutdown
2018-05-25T10:04:38.390Z airtap browser error (android@6.0: [get("http://airtap.local:40573/airtap")] Error response status: 13, UnknownError - An unknown server-side error occurred while processing the command. Selenium error: An unknown server-side error occurred while processing the command. Original error: Could not proxy. Proxy error: UiAUtomator shut down unexpectedly), restarting
2018-05-25T10:04:38.393Z airtap:sauce:android:6.0 running
2018-05-25T10:04:38.394Z airtap:setup bouncer active on port 41253
ok 60 browser.on.init called
2018-05-25T10:04:38.394Z airtap:sauce:android:6.0 queuing
2018-05-25T10:05:31.424Z airtap:sauce:android:6.0 open http://airtap.local:41253/airtap
2018-05-25T10:05:32.888Z airtap:control-app zuul client took 0.376 seconds to bundle
2018-05-25T10:05:33.782Z airtap:browserify test files took 0.089 seconds to bundle
2018-05-25T10:05:36.657Z airtap:sauce:android:6.0 browser opened url
2018-05-25T10:05:36.658Z airtap:sauce:android:6.0 waiting for test results from http://airtap.local:41253/airtap
2018-05-25T10:05:37.394Z airtap:sauce:android:6.0 res.length: 100
2018-05-25T10:05:37.395Z airtap:sauce:android:6.0 done
2018-05-25T10:05:37.396Z airtap:sauce:android:6.0 finished tests for http://airtap.local:41253/airtap
2018-05-25T10:05:37.658Z airtap:sauce:android:6.0 quitting browser
2018-05-25T10:05:38.350Z airtap:sauce:android:6.0 shutdown
ok 61 should be equivalent
ok 62 should be equivalent
ok 63 should be equivalent
ok 64 should be equivalent
ok 65 should be equivalent
ok 66 should be equal
ok 67 should be equal
```

**Comment by vweevers** [comment]
Questions:

Is this called on retry?

https://github.com/airtap/airtap/blob/e1626abe600f3787197dfb9cb7dd0bad9d9bfc87/test/integration/verify-common.js#L15

Is this called before retry?

https://github.com/airtap/airtap/blob/e1626abe600f3787197dfb9cb7dd0bad9d9bfc87/test/integration/verify-common.js#L36

**Comment by ralphtheninja** [comment]
> Maybe we can mock Sauce Labs (webdriver) and write unit tests for failing (and restarting) browsers.

We should probably do this. I think the `.shutdown()` logic in `sauce-browser.js` is pretty hard to follow and I'm wondering if we can't simplify this logic quite a bit, this together with the start logic in `lib/airtap.js` is a minor disaster :wink: 

Just take a look at the startup logic:

https://github.com/airtap/airtap/blob/e1626abe600f3787197dfb9cb7dd0bad9d9bfc87/lib/airtap.js#L107-L133

It's a mishmash of events and error events and callbacks.

`t.pass('browser.on.init called')` is called once `createTestServer()` _succeeds_:

https://github.com/airtap/airtap/blob/e1626abe600f3787197dfb9cb7dd0bad9d9bfc87/lib/sauce-browser.js#L42-L46

So the error we get above happens _before_ we emit `'init'`.

Ideally, I'd like to have something like this:

```js
browser.start(function (err, stats) {
  if (err) { // someting went wrong, restart? }
  else { // successful, use stats }
})
```

**Comment by vweevers** [comment]
> Ideally, I'd like to have something like this:
> ```js
> browser.start(function (err, stats) {
>  if (err) { // someting went wrong, restart? }
>  else { // successful, use stats }
> })
> ```

Def. I started work on that earlier (and got to a pretty nice point) but it's now too old to pick up again.

**Comment by ralphtheninja** [comment]
To summarize, I think the only way we can get rid of shaky tests, is to re-write the whole startup/shutdown logic. Right now it's just too messy and difficult to understand.

**Comment by ralphtheninja** [comment]
> Def. I started work on that earlier (and got to a pretty nice point) but it's now too old to pick up again.

We can work on this together and brain storm a bit. Let me know when you have time and feel like it.

**Comment by vweevers** [comment]
Tests are stable (enough) now.
